embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/arch/DataSourceState.kt

```diff
@@ -0,0 +1,68 @@
+package io.embrace.android.embracesdk.arch
+
+/**
+ * Holds the current state of the service. This class automatically handles changes in config
+ * that enable/disable the service, and creates new instances of the service as required.
+ * It also is capable of disabling the service if the [EnvelopeType] is not supported.
+ */
+internal class DataSourceState<T : DataSource<R>, R>(
+
+    /**
+     * Provides instances of services. A service must define an interface
+     * that extends [DataSource] for orchestration. This helps enforce testability
+     * by making it impossible to register data capture without defining a testable interface.
+     */
+    factory: () -> DataSource<R>,
+
+    /**
+     * Predicate that determines if the service should be enabled or not, via a config value.
+     */
+    private val configGate: () -> Boolean,
+
+    /**
+     * The type of envelope that contains the data.
+     */
+    private var currentEnvelope: EnvelopeType,
```

Comment:
We should come up with a new term to the thing that is either a session or a background activity. We use the term envelope to describe the thing we stuff custom payloads into.

Calling it a Session makes the most sense, frankly, but we have it so ingrained that a background activity is not a session that this may be hard. That said, both smell like Sessions to me and probably to a lot of people in the industry - so is this when we bite the bullet and just call this that, and have two types - foreground and background?

This probably requires a bigger discussion than just this review.

Reply:
Agree - I will ask the android guild for some suggestions